File: app/src/main/java/com/dougkeen/bart/activities/AbstractViewActivity.java
package com.dougkeen.bart.activities;

import androidx.appcompat.app.AppCompatActivity;

import com.dougkeen.bart.BartRunnerApplication;

import java.util.concurrent.TimeUnit;

public abstract class AbstractViewActivity extends AppCompatActivity {

    private static final int MAXIMUM_IDLE_HOURS = 3;

    @Override
    protected void onStart() {
        super.onStart();
        BartRunnerApplication application = (BartRunnerApplication) getApplication();
        long lastActivity = application.getActivityTimestamp();
        long currentTime = System.currentTimeMillis();
        long timeDifference = currentTime - lastActivity;
        if (lastActivity > 0
                && TimeUnit.MILLISECONDS.toHours(timeDifference) >= MAXIMUM_IDLE_HOURS) {
            finish();
        }
    }
}

File: app/src/main/kotlin/com/dougkeen/bart/activities/AbstractViewActivity.kt
package com.dougkeen.bart.activities

import androidx.appcompat.app.AppCompatActivity
import com.dougkeen.bart.BartRunnerApplication
import com.dougkeen.bart.model.TimeSource
import java.util.concurrent.TimeUnit

abstract class AbstractViewActivity : AppCompatActivity() {
    protected fun activityTimeSource(): TimeSource =
        (application as BartRunnerApplication).timeSource

    override fun onStart() {
        super.onStart()
        val application = application as BartRunnerApplication
        val lastActivity = application.activityTimestamp
        val timeDifference = activityTimeSource().nowMillis() - lastActivity
        if (lastActivity > 0
            && TimeUnit.MILLISECONDS.toHours(timeDifference) >= MAXIMUM_IDLE_HOURS
        ) {
            finish()
        }
    }

    private companion object {
        const val MAXIMUM_IDLE_HOURS = 3L
    }
}
